lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -195,23 +109,73 @@ type stage3 = {
   cfg : Cfg.t;
   failures : (addr * error) list;
 }
+
 type t = stage3
 
 let errored s ty = {s with errors = (s.addr,ty) :: s.errors}
 
-
 let update_dests s dests mem =
-  let key = Memory.max_addr mem in
+  let dests = List.map dests ~f:(fun d -> match d with
+      | Some addr,_ when Memory.contains s.base addr -> d
+      | _,kind -> None, kind) in
+  let key = Memory.min_addr mem in
   match dests with
   | [] -> Addr.Table.add_exn s.dests ~key ~data:[]
   | dests -> List.iter dests ~f:(fun data ->
       Addr.Table.add_multi s.dests ~key ~data)
 
+let collect ~dir ~stop_on ?cond s start =
+  let add = match cond with
+    | None -> fun x xs -> x :: xs
+    | Some cond -> fun x xs -> if cond x then x :: xs else xs in
+  let rec loop addr data =
+    match dir s.visited addr with
+    | None -> data
+    | Some x when stop_on x -> data
+    | Some x -> loop (fst x) (add x data) in
+  loop start []
+
+let exists_visited ~dir ~stop_on ~cond s start  =
+  not (List.is_empty @@ collect ~dir ~stop_on ~cond s start)
+
+let is_shared_max (_,a) (_,a') = Addr.equal a a'
+
+let need_barrier s ((addr,max_addr) as insn) =
```

Comment:
it is not a barrier, but a _terminator_ (barrier is basically an instruction that is followed by data). 

could you please explain all the clauses? 

Also, it is totally inefficient to construct the list,  just to prove that it is not empty. Please, get rid of it.